File: recursion/search_range.js
var searchRange = function(nums, target) {
  if (nums.length === 1 && nums[0] === target) {
    return [0, 0];
  }
  var left = 0;
  var right = nums.length - 1;
  var begin = binaryEdgeSearch(nums, left, right, target, -1);
  console.log(begin)
  var end = nums.length - 1 - binaryEdgeSearch(nums.reverse, left, right, target, -1);

  if (begin >= 0 && end >= 0) {
    return [begin, end];
  } else {
    return [-1, -1];
  }
};

var binaryEdgeSearch = function(nums, left, right, target, edgeIncrement) {
  if (left + 1 === right) {
    if (nums[left] === target) {
      return left;
    }
    if (nums[right] === target) {
      return right;
    }
    return null;
  }

  var mid = Math.floor(left + (right - left) / 2);

  if (nums[mid] === target) {
    if (nums[mid + edgeIncrement] != target) {
      return mid;
    } else {
      right = mid + edgeIncrement;
    }
  } else if (nums[mid] < target) {
    left = mid + 1;
  } else {
    right = mid - 1;
  } 

  binaryEdgeSearch(nums, left, right, target, edgeIncrement);
};

console.log(searchRange([5,7,7,8,8,10], 8));

File: recursion/search_range.rb
require 'pry'
# Given an array of integers sorted in ascending order, find the starting and ending position of a given target value.
# 
# Your algorithm's runtime complexity must be in the order of O(log n).
# 
# If the target is not found in the array, return [-1, -1].
# 
# For example,
# Given [5, 7, 7, 8, 8, 10] and target value 8,
# return [3, 4].
#
# binary search recursive helper for beginning of range
# another bsearch helper for end
# 
#

def search_range(nums, target)
  range_begin = find_beginning(nums, target, 0, nums.length - 1)
  range_end = find_ending(nums, target, 0, nums.length - 1)

  [range_begin, range_end]
end

def find_beginning(nums, target, left, right)
  if left + 1 >= right
    return left if nums[left] == target
    return right if nums[right] == target
    return -1
  end

  mid = left + (right - left) / 2

  if nums[mid] >= target
    right = mid
  else
    left = mid
  end

  find_beginning(nums, target, left, right)
end

def find_ending(nums, target, left, right)
  if left + 1 >= right
    return right if nums[right] == target
    return left if nums[left] == target
    return -1
  end

  mid = left + (right - left) / 2

  if nums[mid] <= target
    left = mid
  else
    right = mid
  end

  find_ending(nums, target, left, right)
end

puts search_range([5, 7, 7, 8, 8, 10], 8) == [3, 4]

